Basic forms

rodya-mirov opened the pull request:
Implement the first segment of forms for critique. This has the following:
- Form pages are defined as .js files which are essentially json blobs with a constructor (we'll move to json as soon as we are confident in a file format).
- Form pages have a title, a bit of generic help text, and a series of inputs.
- Each form input has a label, a type (which in some cases includes choices, e.g. a radio button), and a bit of specific help text.

The form pages are loaded in a canonical order (specified in code; this can also be json-ified at some point). The relevant parts are loaded by the InstructionPanel, CenterForm, and ProgressPanel, and basic information is used for navigation, etc. 

The actual content panels are almost entirely data driven, with very little text hard-coded in the component, so no code changes are required to add new content.

NOTE: I'm not good enough at javascript to figure out how to make package-lock.json not generate thousands of lines of garbage json each time. If it's important to you please fix.

Bachmann1234 commented:
> [![Coverage Status](https://camo.githubusercontent.com/b5affc511d81920d482da0167331abb1c801cff6/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f32313037333037372f6261646765)](https://coveralls.io/builds/21073077)
> 
> Coverage remained the same at 89.899% when pulling **[95829be](https://github.com/diy-asylum/formfiller/commit/95829bea76e47292ed7d186d86cae4c0d12e37b5) on basic-forms** into **[194c76d](https://github.com/diy-asylum/formfiller/commit/194c76d27ebbe949dec178432de15a3dc1463438) on master**.

I Might need to hide these since they are a bit misleading since its only reporting backend coverage

Bachmann1234 commented:
You planing on adding any tests to this?

JeuelyFish commented on diy_asylum_frontend/src/App.js:
```diff
@@ -1,5 +1,5 @@
 import React, { Component } from "react";
-import { BrowserRouter as Router, Route } from "react-router-dom";
+import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
```
👌

JeuelyFish commented on diy_asylum_frontend/src/reducers/helpText.js:
```diff
@@ -0,0 +1,42 @@
+import { INCREMENT_STEP, DECREMENT_STEP } from "./form";
```
I think I see where you're going with this. The idea is to reset the helper form to blank whenever the user goes forwards or back in step forms, right?

That's a good edge case to cover, but I think that mixing actions in the reducers (at this stage) is a little much. Let's either, in the component level, explicitly call each action *or* , if we really do what to mix the actions, import only `setInputHelpText` into the form reducer

since it's just one import instead of two

rodya-mirov replied:
I'll just make the next page buttons fire off two actions, seems innocuous. I tend to want to put more logic in the reducers and less in the view, but I guess the common convention is the opposite.

JeuelyFish commented on diy_asylum_frontend/src/components/FormContentComponent/index.js:
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+
+const inputMaker = ({ inputType, elementName, help }, helpTextSetter) => {
```
I like this! This was what I was hoping to make when I briefly mentioned it in this slack message (https://diyasylum.slack.com/archives/CEPM8HDAL/p1546537263006700)

Why don't we just make these there own react components and use them explicitly tho?

It's elegant and abstracted logic as it stands right now, but it we should try and keep as close to the metal as possible in the FE.

JeuelyFish commented:
This is headed in the right direction for sure. I'd suggest moving all of the contents of `src/contentpages` into `src/components/CenterForm` since they are really only used in that component and the word "pages" is a little bit of a misnomer. 

Then change each of those "pages" into "formSections", or whatever name you want, and use the React component constructor. In general, you're abstracting the rendering away from the logic composition... which is neat but it's not going against the grain of the react conventions. Most, if not all, logic should be in either the function or class component form (https://reactjs.org/docs/components-and-props.html#function-and-class-components)

rodya-mirov commented:
@JeuelyFish that bug with the shared form state is very weird. I've put up a stackoverflow post about it because I couldn't figure it out. https://stackoverflow.com/questions/54204288/distinct-react-forms-are-sharing-state-how-to-separate-them

As for the rest of the comments, I've moved all the "functions that make react components" into just react components. I also moved the "content form makers" into the center form directory and made them helpers (and renamed them).

Your statement about `contentPages` was incorrect; that data source is used by all three content panes of the application, not just the center form. It's just the data that drives the application. In fact my hope is that we make it into an external json blob instead of code, so that we can have it translated, nitpicked, and so on.

rodya-mirov commented:
@JeuelyFish Now that I think about it the "right" way for the form state issue to be resolved is to make this into a Controlled Component (probably with Formik), which should solve the issue as a side effect (since we'll manage the state ourselves, instead of trying to force the "hello world" form to solve itself).

I was hoping to do that in a future PR but no time like the present I guess.

mkreisel commented:
Fwiw, I don't think things need to be perfect for us to merge. I think we should document the known issues and regroup to determine a path forward.
